File: Top 100 Codes/03-Codes for Recursion/09-Print All Permutations of a String-/PrintAllPermutationsofaString.py
def permute(s, l, r):
    if l == r:
        print("".join(s))
        return
    
    for i in range(l, r + 1):
        s[l], s[i] = s[i], s[l]  # Swap characters
        permute(s, l + 1, r)  # Recurse for the next character
        s[l], s[i] = s[i], s[l]  # Backtrack to the original order

# Taking input
s = input("Enter a string: ")
print("All permutations of the string:")
permute(list(s), 0, len(s) - 1)

File: Top 100 Codes/03-Codes for Recursion/09-Print All Permutations of a String-/PrintAllPermutationsofaString.cpp
#include <iostream>
#include <string>
using namespace std;

void permute(string s, int l, int r) {
    if (l == r) {
        cout << s << endl;
        return;
    }
    
    for (int i = l; i <= r; i++) {
        swap(s[l], s[i]);  // Swap characters
        permute(s, l + 1, r);  // Recurse for the next character
        swap(s[l], s[i]);  // Backtrack to the original string
    }
}

int main() {
    string str;
    cout << "Enter a string: ";
    cin >> str;
    
    cout << "All permutations of the string:\n";
    permute(str, 0, str.size() - 1);
    
    return 0;
}
